prover_rust/src/prover.rs

```diff
@@ -0,0 +1,324 @@
+use once_cell::sync::Lazy;
+use std::{cell::RefCell, cmp::Ordering, env, rc::Rc};
+
+use crate::{
+    config::Config,
+    coordinator_client::{
+        listener::Listener, types::*, Config as CoordinatorConfig, CoordinatorClient,
+    },
+    geth_client::{types::get_block_number, GethClient},
+    key_signer::KeySigner,
+    types::{CommonHash, ProofFailureType, ProofStatus, ProofType},
+    zk_circuits_handler::{CircuitsHandler, CircuitsHandlerProvider},
+};
+
+use super::types::{ProofDetail, Task};
+use prover::{BlockTrace, ChunkHash, ChunkProof};
+
+// Only used for debugging.
+pub(crate) static OUTPUT_DIR: Lazy<Option<String>> =
+    Lazy::new(|| env::var("PROVER_OUTPUT_DIR").ok());
+
+pub struct Prover<'a> {
+    config: &'a Config,
+    key_signer: Rc<KeySigner>,
+    circuits_handler_provider: CircuitsHandlerProvider,
+    coordinator_client: RefCell<CoordinatorClient>,
+    geth_client: Option<RefCell<GethClient>>,
+}
+
+// a u64 is positive when it's 63th index bit not set
+fn is_positive(n: &U64) -> bool {
+    !n.bit(63)
+}
+
+impl<'a> Prover<'a> {
+    pub fn new(config: &'a Config, coordinator_listener: Box<dyn Listener>) -> Result<Self> {
+        let proof_type = config.core.proof_type;
+        let params_path = &config.core.params_path;
+        let assets_path = &config.core.assets_path;
+        let keystore_path = &config.keystore_path;
+        let keystore_password = &config.keystore_password;
+
+        let coordinator_config = CoordinatorConfig {
+            endpoint: config.coordinator.base_url.clone(),
+            prover_name: config.prover_name.clone(),
+            prover_version: crate::version::get_version(),
+            hard_fork_name: config.hard_fork_name.clone(),
+        };
+
+        let key_signer = Rc::new(KeySigner::new(&keystore_path, &keystore_password)?);
+        let coordinator_client = CoordinatorClient::new(
+            coordinator_config,
+            Rc::clone(&key_signer),
+            coordinator_listener,
+        )?;
+
+        let mut prover = Prover {
+            config,
+            key_signer: Rc::clone(&key_signer),
+            circuits_handler_provider: CircuitsHandlerProvider::new(
+                proof_type,
+                params_path,
+                assets_path,
+            )?,
+            coordinator_client: RefCell::new(coordinator_client),
+            geth_client: None,
+        };
+
+        if config.core.proof_type == ProofType::ProofTypeChunk {
+            prover.geth_client = Some(RefCell::new(GethClient::new(
+                "test",
+                &config.l2geth.as_ref().unwrap().endpoint,
+            )?));
+        }
+
+        Ok(prover)
+    }
+
+    pub fn get_proof_type(&self) -> ProofType {
+        self.config.core.proof_type
+    }
+
+    pub fn get_public_key(&self) -> String {
+        self.key_signer.get_public_key()
+    }
+
+    pub fn fetch_task(&self) -> Result<Task> {
+        let vks = self.circuits_handler_provider.get_vks();
```

Comment:
why get the vk every time, can we store it when starts?

Reply:
done